core/src/Cache/CacheFallbackOperations.cs

```diff
@@ -96,6 +102,14 @@ internal class CacheFallbackOperations
                     return;
                 }
 
+                Debug.Assert(
```

Comment:
These should only be hit in tests

Reply:
not convinced

---
In reply to: [219185322](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1259#discussion_r219185322) [](ancestors = 219185322)

Reply:
Don't we also run our tests in RELEASE mode?  If so, these won't be hit in test code, and then what?

Reply:
No, tests run in Debug mode. Also, when we perform manual testing using the dev apps, we use the debug version of the library. 

The reason for using Debug.Assert is that I am not sure there aren't cases where those conditions don't actually hold. The implications in this case are minimal for the user (e.g. minor loss of SSO), so I am hesitant to crash the app with an exception. I am also hesitant to add a lot of validation code in hot paths of the library when it is used in production.

Reply:
IMHO, we should run tests in RELEASE mode.  That's where our customers are using us and that's what we should be testing.  Testing in debug mode should be "yup, we do that too for sanity" but we should be testing what customers are running.

I'll need to look more at this particular scenario for what you're trying to achieve with the asserts, but we should be able to find a way to validate debug/release equally without un-naturally burdening users with exceptions/errors that aren't errors/exceptions.

Reply:
I would prefer we identify a way to test with-out Debug.Asserts

---
In reply to: [219557012](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1259#discussion_r219557012) [](ancestors = 219557012)

Reply:
There is no functional difference between DEBUG and RELEASE modes and .net provides ample support for adding Debug only code. Agreed that we should never perf test against Release. 

VS logic uses this technique a lot, especially to protect against changes from other parts of the app. 

it's also a way of "documenting" the code by identifying invariants. It took me a long time to realize that those conditions always hold true.

Reply:
How about logging?

---
In reply to: [219556166](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1259#discussion_r219556166) [](ancestors = 219556166)

Reply:
Discussed offline, will do logging and bring this back at another time.

Reply:
Have we seen any issues like this ever occur? Is it something critical? Can we consider only doing this validation in case there is not token returned or an error happens?

---
In reply to: [220294785](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1259#discussion_r220294785) [](ancestors = 220294785)